File: solutions/go/knapsack/1/knapsack.go
package knapsack

type Item struct {
	Weight, Value int
}

// Knapsack takes in a maximum carrying capacity and a collection of items
// and returns the maximum value that can be carried by the knapsack
// given that the knapsack can only carry a maximum weight given by maximumWeight
func Knapsack(maxWeight int, items []Item) int {
	return knapsack(maxWeight, items, map[index]int{})
}

type index struct {
	nElems    int
	maxWeight int
}

func knapsack(maxWeigth int, items []Item, mem map[index]int) int {
	if maxWeigth == 0 || len(items) == 0 {
		return 0
	}
	if sol, ok := mem[index{len(items), maxWeigth}]; ok {
		return sol
	}
	lastItem := items[len(items)-1]
	var sol int
	if lastItem.Weight <= maxWeigth {
		sol = max(
			lastItem.Value+knapsack(maxWeigth-lastItem.Weight, items[:len(items)-1], mem),
			knapsack(maxWeigth, items[:len(items)-1], mem))
	} else {
		sol = knapsack(maxWeigth, items[:len(items)-1], mem)
	}
	mem[index{len(items), maxWeigth}] = sol
	return sol

}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: solutions/elixir/knapsack/1/lib/knapsack.ex
defmodule Knapsack do
  @doc """
  Return the maximum value that a knapsack can carry.
  """
  @spec maximum_value(items :: [%{value: integer, weight: integer}], maximum_weight :: integer) ::
          integer

  def maximum_value(items, maximum_weight) do
    knapsack(items, maximum_weight, length(items))
  end

  defp value_of(items, n), do: Enum.at(items, n)[:value]

  defp weight_of(items, n), do: Enum.at(items, n)[:weight]

  defp knapsack(_, 0, _), do: 0

  defp knapsack(_, _, 0), do: 0

  defp knapsack(items, max_weight, n) do
    if weight_of(items, n-1) > max_weight do
      knapsack(items, max_weight, n-1)
    else
      max(
        value_of(items, n-1) + knapsack(items, max_weight - weight_of(items, n-1), n-1),
        knapsack(items, max_weight, n-1))
    end
  end

end
